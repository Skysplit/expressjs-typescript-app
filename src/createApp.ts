import express from 'express';
import { urlencoded, json } from 'body-parser';
import compression from 'compression';
import cookieParser from 'cookie-parser';
import routes from './routes';

const createApp = async () => {
  const app = express();

  // App settings
  app.disable('x-powered-by');

  // Global middleware
  app.use(compression());
  app.use(cookieParser());
  app.use(json());
  app.use(urlencoded({ extended: true }));

  // Mount routes
  app.use(routes);

  return app;
};

export default createApp;
