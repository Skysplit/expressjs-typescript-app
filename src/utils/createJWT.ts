import jwt from 'jsonwebtoken';
import appSecret from './appSecret';

export default (payload: any, options?: jwt.SignOptions) => jwt.sign(payload, appSecret, options);
