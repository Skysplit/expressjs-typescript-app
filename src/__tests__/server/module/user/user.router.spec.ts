import request, { SuperTest, Test } from 'supertest';
import { Application } from 'express';
import createApp from '@server/createApp';
import User from '@server/module/user/user.model';
import createJWT from '@server/utils/createJWT';

describe('/users', () => {
  let app: Application;
  let user: User;
  let req: SuperTest<Test>;

  beforeAll(async () => {
    app = await createApp();

    user = User.create({
      email: 'test@example.com',
      password: 'Secret99',
    });

    await user.save();
  });

  beforeEach(async () => {
    req = request(app);
  });

  describe('/me#GET', () => {
    test('should access user profile', async () => {
      const token = createJWT(user.toJSON());
      const response = await req.get('/users/me')
        .set({ Authorization: `Bearer ${token}` });

      expect(response.status).toEqual(200);
      expect(response.body).toEqual(
        JSON.parse(JSON.stringify(user)),
      );
    });

    describe('when no bearer token is provided', () => {
      test('should forbid profile access', async () => {
        const response = await req.get('/users/me');

        expect(response.status).toEqual(401);
      });
    });
  });

  describe('/login#POST', () => {
    test('should respond with JWT token', async () => {
      const response = await req.post('/users/login').send({
        email: 'test@example.com',
        password: 'Secret99',
      });

      expect(response.body).toEqual({
        success: true,
        token: expect.any(String),
      });
      expect(response.status).toEqual(200);
    });

    describe('when no credentials are provided', () => {
      test('should display missing credentials', async () => {
        const response = await req.post('/users/login');
        expect(response.status).toEqual(422);
        expect(response.body).toMatchSnapshot();
      });
    });

    describe('when wrong password is provided', () => {
      test('should display wrong credentials', async () => {
        const response = await req.post('/users/login').send({
          email: 'test@example.com',
          password: 'Secret1',
        });
        expect(response.status).toEqual(422);
        expect(response.body).toMatchSnapshot();
      });
    });

    describe('when wrong email is provided', () => {
      test('should display wrong credentials', async () => {
        const response = await req.post('/users/login').send({
          email: 'wrong@example.com',
          password: 'Secret99',
        });
        expect(response.status).toEqual(422);
        expect(response.body).toMatchSnapshot();
      });
    });
  });

});
