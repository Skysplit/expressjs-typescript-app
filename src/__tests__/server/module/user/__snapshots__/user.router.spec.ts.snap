// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/users /login#POST when no credentials are provided should display missing credentials 1`] = `
Object {
  "message": "Missing credentials",
}
`;

exports[`/users /login#POST when wrong email is provided should display wrong credentials 1`] = `
Object {
  "message": "Incorrect credentials",
}
`;

exports[`/users /login#POST when wrong password is provided should display wrong credentials 1`] = `
Object {
  "message": "Incorrect credentials",
}
`;
